src/mpi/attr/attr_impl.c

```diff
@@ -594,15 +597,15 @@ int MPIR_Type_set_attr_impl(MPIR_Datatype * type_ptr, MPII_Keyval * keyval_ptr,
     while (p) {
         if (p->keyval->handle == keyval_ptr->handle) {
             /* If found, call the delete function before replacing the
-             * attribute */
+             * attribute. If the delete function fails on the old
+             * attribute value, we keep the old value. */
             mpi_errno = MPIR_Call_attr_delete(type_ptr->handle, p);
-            /* --BEGIN ERROR HANDLING-- */
-            if (mpi_errno) {
+            if (mpi_errno != MPI_SUCCESS) {
                 goto fn_fail;
             }
-            /* --END ERROR HANDLING-- */
-            p->value = (MPII_Attr_val_t) (intptr_t) attribute_val;
             p->attrType = attrType;
+            p->value = (MPII_Attr_val_t) (intptr_t) attribute_val;
+            /* Does not change the reference count on the keyval */
             break;
         } else if (p->keyval->handle > keyval_ptr->handle) {
```

Comment:
@hzhou Can you please check this part of the code for datatype (and similarly for windows)? I cannot make sense why this branch is needed, and why things are not done the same way for communicators.

It looks like it is somehow related to wanting to insert in a keyval order in the list rather than in the tail . I have no objections to that, really, but I found the asymmetry in communicators striking. I noticed the comment /* ``/* CHANGE FOR MPI 2.2: If not found, add at the beginning */``, but I don't remember any specific standard mandate about that.

Reply:
All the attributes are linearly ordered in linked list. This branch is where we found the position where the new entry need be inserted. Now that I am looking at this code, seems there is a bug that we are inserting after `p` but seems we should insert before `p`. 

Apparently, the comm path was changed at some point and we neglected to apply the changes to Type and Win. I don't know the original reason for the change. I'll follow up.

Reply:
Found the original commit:
```
commit 8f959cd7bf79dd335a0d5d439643bafa4080677e (HEAD)
Author: David Goodell <goodell@mcs.anl.gov>
Date:   Thu Aug 20 20:45:04 2009 +0000

    [svn-r5250] Attribute delete callbacks are now ordered (MPI-2.2 ticket #71).

    Implementation of MPI-2.2 ticket #71.  Delete callbacks are now invoked
    in LIFO order.  Once the MPI version number macros are updated to 2.2
    the test added in r4346 will check for this behavior.  This change also
    includes another test for LIFO ordering because I wrote it before I saw
    Bill's original test.

    For more information and history on this issue, please see this URL:

    https://svn.mpi-forum.org/trac/mpi-forum-web/ticket/71

    The non-testing portions of this change were originally written by Bill Gropp.

 src/mpi/attr/comm_set_attr.c |  24 +++++---------------
 test/mpi/attr/Makefile.sm    |   1 +
 test/mpi/attr/attrend2.c     | 128 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 test/mpi/attr/testlist       |   2 ++
 4 files changed, 137 insertions(+), 18 deletions(-)
```

Reply:
The comment in the test is clearer:
```
/*
      The MPI-2.2 specification makes it clear that attributes are called on
      MPI_COMM_WORLD and MPI_COMM_SELF at the very beginning of MPI_Finalize in
      LIFO order with respect to the order in which they are set.  This is
      useful for tools that want to perform the MPI equivalent of an "at_exit"
      action.
 */
```
I guess that is the reason the change was only applied to the comm, but not type or win. I think we should update the type and win as well to make it consistent.

Reply:
That comment is a inaccurate, though. The MPI 2.2 standard only talks about such ordering for `MPI_COMM_SELF` only, and that's still the case in MPI-4.1.
Anyway, I think your intention of making everything consistent makes sense, despite the small 0.5 performance penalty on the average for objects with lots and lots of attributes (with should be an extremely rare case).
Alternatively, you could just special case `MPI_COMM_SELF` and skip the branch to force the LIFO order only for that case. I think I still prefer LIFO everywhere, just my two cents.